ddtrace/contrib/internal/langchain/utils.py

```diff
@@ -0,0 +1,125 @@
+            return chunk
+        except StopIteration:
+            self._on_span_finish(self._dd_span, self._chunks)
+            self._dd_span.finish()
+            raise
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            self._dd_integration.metric(self._dd_span, "incr", "request.error", 1)
+            raise
+
+
+class TracedLangchainAsyncStreamResponse(BaseTracedLangChainStreamResponse):
+    async def __aenter__(self):
+        await self._generator.__enter__()
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self._generator.__exit__(exc_type, exc_val, exc_tb)
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        try:
+            chunk = await self._generator.__anext__()
+            self._chunks.append(chunk)
+            return chunk
+        except StopAsyncIteration:
+            self._on_span_finish(self._dd_span, self._chunks)
+            self._dd_span.finish()
+            raise
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            self._dd_integration.metric(self._dd_span, "incr", "request.error", 1)
+            raise
+
+
+def shared_stream(
+    integration,
+    pin,
+    func,
+    instance,
+    args,
+    kwargs,
+    interface_type,
+    on_span_started,
+    on_span_finished,
+    **extra_options,
+):
+    options = {
+        "pin": pin,
+        "operation_id": f"{instance.__module__}.{instance.__class__.__name__}",
+        "interface_type": interface_type,
+    }
+
+    options.update(extra_options)
+
+    span = integration.trace(**options)
+    span.set_tag("langchain.request.stream", True)
+    on_span_started(span)
+
+    try:
+        resp = func(*args, **kwargs)
+        cls = TracedLangchainAsyncStreamResponse if inspect.isasyncgen(resp) else TracedLangchainStreamResponse
+
+        return cls(resp, integration, span, on_span_finished)
+    except Exception:
+        # error with the method call itself
+        span.set_exc_info(*sys.exc_info())
+        span.finish()
+        integration.metric(span, "incr", "request.error", 1)
+        integration.metric(span, "dist", "request.duration", span.duration_ns)
+        raise
+
+
+def tag_general_message_input(span, inputs, integration, langchain_core):
+    if langchain_core and isinstance(inputs, langchain_core.prompt_values.PromptValue):
+        inputs = inputs.to_messages()
+    elif not isinstance(inputs, list):
+        inputs = [inputs]
+    for input_idx, inp in enumerate(inputs):
+        if isinstance(inp, dict):
+            span.set_tag_str(
+                "langchain.request.messages.%d.content" % (input_idx),
+                integration.trunc(str(inp.get("content", ""))),
+            )
+            span.set_tag_str(
+                "langchain.request.messages.%d.role" % (input_idx),
+                str(inp.get("role", "")),
+            )
```

Comment:
Let's only attach role if role is not None, in case of llm.stream() being called rather than chat_model.